token-metadata/js/src/custom/metadata-deserializer.ts

```diff
@@ -0,0 +1,96 @@
+import * as beet from '@metaplex-foundation/beet';
+import * as beetSolana from '@metaplex-foundation/beet-solana';
+import { Metadata } from '../generated/accounts/Metadata';
+import { collectionBeet } from '../generated/types/Collection';
+import { dataBeet } from '../generated/types/Data';
+import { keyBeet } from '../generated/types/Key';
+import { tokenStandardBeet } from '../generated/types/TokenStandard';
+import { usesBeet } from '../generated/types/Uses';
+
+/**
+ * This is a custom deserializer for TokenMetadata in order to mitigate acounts with corrupted
+ * data on chain.
+ *
+ * Instead of failing the deserialization for the section that is possibly corrupt it just returns
+ * `null` for the fields that would normally be stored in that section.
+ *
+ * This deserializer matches the [fix implemented in the Rust program](https://github.com/metaplex-foundation/metaplex-program-library/blob/df36da5a78fb17e1690247b8041b761d27c83b1b/token-metadata/program/src/deser.rs#L6).
+ * Also @see ../../../program/src/deser.rs
+ */
+export function deserialize(buf: Buffer, offset = 0): [Metadata, number] {
+  let cursor = offset;
+
+  // key
+  const key = keyBeet.read(buf, cursor);
+  cursor += keyBeet.byteSize;
+
+  // updateAuthority
+  const updateAuthority = beetSolana.publicKey.read(buf, cursor);
+  cursor += beetSolana.publicKey.byteSize;
+
+  // mint
+  const mint = beetSolana.publicKey.read(buf, cursor);
+  cursor += beetSolana.publicKey.byteSize;
+
+  // data
+  const [data, dataDelta] = dataBeet.deserialize(buf, cursor);
+  cursor = dataDelta;
+
+  // primarySaleHappened
+  const primarySaleHappened = beet.bool.read(buf, cursor);
+  cursor += beet.bool.byteSize;
+
+  // isMutable
+  const isMutable = beet.bool.read(buf, cursor);
+  cursor += beet.bool.byteSize;
+
+  // editionNonce
+  const editionNonceBeet = beet.coption(beet.u8).toFixedFromData(buf, cursor);
+  const editionNonce = editionNonceBeet.read(buf, cursor);
+  cursor += editionNonceBeet.byteSize;
+
+  // -----------------
+  // Possibly corrupted section
+  // -----------------
+
+  // tokenStandard
+  const [tokenStandard, tokenDelta] = tryReadOption(beet.coption(tokenStandardBeet), buf, cursor);
+  cursor += tokenDelta;
+
+  // collection
+  const [collection, collectionDelta] = tryReadOption(beet.coption(collectionBeet), buf, cursor);
+  cursor += collectionDelta;
+
+  // uses
+  const [uses, usesDelta] = tryReadOption(beet.coption(usesBeet), buf, cursor);
+  cursor += usesDelta;
+
+  const args = {
+    key,
+    updateAuthority,
+    mint,
+    data,
+    primarySaleHappened,
+    isMutable,
+    editionNonce,
+    tokenStandard,
+    collection,
+    uses,
+  };
+
+  return [Metadata.fromArgs(args), cursor];
+}
+
+function tryReadOption<T>(
```

Comment:
Note that this doesn't exactly match the behavior of the Rust deserializer. There we check TokenStandard, Collection and Uses and set *all* to `None` if *any* of them are corrupted. This is because the corruption is likely to be all-or-nothing and it's theoretically possible for some of these fields to have accidentally valid, but logically incorrect, values.

Reply:
That's what I did try to clarify previously, i.e. could one field be corrupted but another not.

If this is important then I can adapt this, but afaik in the case you're describing we'd get `None` for all of them anyways.

Please LMK though if this is a blocker to merge.

Reply:
I updated the PR to have the same behavior as the Rust program.
https://github.com/metaplex-foundation/metaplex-program-library/pull/455/commits/35d12e0f841b6eb042bf146075e5fa7db49f394c